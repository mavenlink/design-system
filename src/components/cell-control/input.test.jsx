import React, { createRef } from 'react';
import { render as _render, screen } from '@testing-library/react';
import user from '@testing-library/user-event';
import Input from './input.jsx';

const render = (ui, options = { labelledBy: 'labelled-by' }) => (
  _render(ui, {
    ...options,
    wrapper: props => (
      <table>
        <thead>
          <tr>
            <th id={options.labelledBy}>Column Header</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            {props.children}
          </tr>
        </tbody>
      </table>
    ),
  })
);

describe('Input cell control', () => {
  const requiredProps = {
    id: 'test-id',
    labelledBy: 'labelled-by',
    name: 'test-name',
  };

  it('has defaults', () => {
    const ref = createRef();
    render(<Input {...requiredProps} ref={ref} />);
    expect(document.body).toMatchSnapshot();
    expect(document.body).toBe(document.activeElement);
    expect(ref.current).toMatchSnapshot();
  });

  it('className API', () => {
    render(<Input {...requiredProps} className="unique-class-name" />);
    expect(document.body).toMatchSnapshot();
  });

  describe('id API', () => {
    it('is a string', () => {
      render(<Input {...requiredProps} id="unique-id" />);
      expect(document.body).toMatchSnapshot();
    });
  });

  describe('labelledBy API', () => {
    it('is a string', () => {
      render(<Input {...requiredProps} labelledBy="unique-labelledby" />, { labelledBy: 'unique-labelledby' });
      expect(screen.getByRole('gridcell')).toHaveAttribute('aria-labelledby', 'unique-labelledby');
    });
  });

  describe('name API', () => {
    it('is a string', () => {
      render(<Input {...requiredProps} name="unique-name" />);
      expect(screen.getByRole('textbox')).toHaveAttribute('name', 'unique-name');
    });
  });

  describe('onChange API', () => {
    it('calls onChange when a new value is selected', () => {
      const onChange = jest.fn(event => event.persist());

      render(<Input {...requiredProps} onChange={onChange} />);

      user.type(screen.getByRole('textbox'), 'hello world!');

      expect(onChange).toHaveBeenCalledWith(expect.objectContaining({
        target: expect.objectContaining({ value: 'hello world!' }),
      }));
    });
  });

  describe('readOnly API', () => {
    it('is true', () => {
      render(<Input {...requiredProps} readOnly={true} />);
      expect(screen.getByRole('textbox')).toHaveAttribute('readonly');
    });

    it('is false', () => {
      render(<Input {...requiredProps} readOnly={false} />);
      expect(screen.getByRole('textbox')).not.toHaveAttribute('readonly');
    });
  });

  describe('required API', () => {
    it('is true', () => {
      render(<Input {...requiredProps} required={true} />);
      expect(screen.getByRole('textbox')).toHaveAttribute('required');
    });

    it('is false', () => {
      render(<Input {...requiredProps} required={false} />);
      expect(screen.getByRole('textbox')).not.toHaveAttribute('required');
    });
  });

  describe('validationMessage API', () => {
    it('is a string', () => {
      render(<Input {...requiredProps} validationMessage="A unique error message." />);
      expect(screen.getByRole('textbox')).toBeInvalid();
      expect(screen.getByRole('textbox')).toHaveDescription('A unique error message.');
    });
  });

  describe('value API', () => {
    it('can be supplied', () => {
      render(<Input {...requiredProps} value="Cheese burrito" />);
      expect(screen.getByRole('textbox')).toHaveValue('Cheese burrito');
    });
  });
});
