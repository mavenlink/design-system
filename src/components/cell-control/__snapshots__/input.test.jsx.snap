// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input cell control has defaults 1`] = `
<body>
  <div>
    <table>
      <thead>
        <tr>
          <th
            id="labelled-by"
          >
            Column Header
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            aria-labelledby="labelled-by"
            class="container"
            role="gridcell"
          >
            <div
              style="position: relative;"
            >
              <input
                aria-describedby="test-idHint "
                class="input"
                id="test-id"
                type="text"
                value=""
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;

exports[`Input cell control has defaults 2`] = `
Object {
  "dirty": false,
  "name": undefined,
  "value": "",
}
`;

exports[`Input cell control id API is a string 1`] = `
<body>
  <div>
    <table>
      <thead>
        <tr>
          <th
            id="labelled-by"
          >
            Column Header
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            aria-labelledby="labelled-by"
            class="container"
            role="gridcell"
          >
            <div
              style="position: relative;"
            >
              <input
                aria-describedby="unique-idHint "
                class="input"
                id="unique-id"
                type="text"
                value=""
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;
